### DISCUSSION TITLE
Support for interfaces, type alias etc

### DISCUSSION THREAD

**Discussion by ahupp** [typescript-rtti/typescript-rtti#3] [Ideas]
I have some interfaces and would like to use reflection to generate type assertions for them:

```
interface Car {
  make: string;
}

function isCar<T>(value: any): assert is Car { ...}
```

It looks like typescript-rtti only generates type info for classes.  Is it possible to do this for interfaces and type aliases as well?

**Comment by rezonant** [comment]
Wow I never saw this, thanks Github!
Anyway @ahupp interfaces are supported now, take another look and file any issues if you run into problems!

**Reply by ahupp** [reply]
It looks like the transformer only handles calls to reflect with a literal type parameter, but not a generic parameter.  So e.g I can't write this function:

```
function assertMatchesShape<T>(val: any) assert val is T {
  return reflect<T>().matchesShape(val);
}
```

Produces:

```
    return reflect<T>().matchesValue(raw)
                  ^
ReferenceError: IΦT is not defined
```

Would you accept a PR to add that function to typescript-rtti?

**Reply by rezonant** [reply]
I'd want to hear exactly how solving that problem would be approached, I think it's not trivial and that various ways of approaching that problem have edge cases that preclude them from being proper solutions.

**Reply by rezonant** [reply]
Right now the blocker is the "opt in" problem. There needs to be a reliable way to detect that the function wants to receive the incoming type information, which cannot be trivially done across separate builds, as the calling code (pkg B) cannot determine if the called function (pkg A) is using reflect<>(), and thus needs the type information

**Reply by rezonant** [reply]
Of course reify<T>() solves this issue by making the interface passable but it requires the caller to participate

**Reply by rezonant** [reply]
OK, I think I've got a solution to this, will update once it's done

**Reply by rezonant** [reply]
OK, version 0.4.0 now supports this but it's *slightly* different for the function implementor to ensure that this works across multiple packages which are each compiled separately by typescript-rtti.

```typescript
import { CallSite, reflect } from 'typescript-rtti';
function assertMatchesShape<T>(val: any, callSite? : CallSite) assert val is T {
  return reflect<T>().matchesShape(val);
}
```

The `callSite` parameter is the opt-in for receiving the metadata. You don't actually have to make use of it in this case unless you want to reflect on the actual parameters (in the future I'd like to add return type and `this` type as well).

**Comment by ahupp** [comment]
Awesome, thanks!
